app/boards/arm/nrfmicro/nrfmicro_11_flipped.yaml

```diff
@@ -1,5 +1,5 @@
 identifier: nrfmicro_11_flipped
-name: nrfmicro_11_flipped
+name: nRFMicro v1.1 Flipped
```

Comment:
The nRFMicro is a prime example of the potential for *families* of versions/revisions and variants, which I'm guessing we'll want to group together at some point.  If we don't account for how we group/link them, it may be error prone.

Thoughts on making them part of the `yaml` structure?
e.g. `nrfmicro.yml`:
```yaml
id: nrfmicro
name: nRFMicro
url: https://github.com/joric/nrfmicro
versions: 
  - id: nrfmicro_11
    name: nRFMicro v1.1
    variants:
    - id: flipped
  - id: nrfmicro_13
    name: nrfMicro v1.3
    version: 1.3
```
I'm just spitballing here.  I'm not even sure if I like the idea myself.

Reply:
I kind of like having the separate yamls for each version. What if we kept the individual one and had a `nrfmicro.yaml` that connects the three existing yamls together as versions. Or I guess the path could be utilized to figure that out too, but I'd rather not rely on the path.

Reply:
The only downside here is complicating the simple version, always requiring two files...

Thoughts on single file, but using an `import` or `include` approach to make it effectively one file, bit allowing individual management?

Reply:
I do like the idea of importing/including. Makes sense for projects that are just called "board" but then have a revision later. In that case, the original can be imported with just changes done to id/anything else.

Reply:
Ok, so digging into upstream Zephyr here a bit, they definitely use a "YAML file per version" for boards, at least. Take a look at `zephyr/boards/arm/nucleo_l452re` for example, that contains both `nucleo_l452re.yaml` and `nucleo_l452re_p.yaml` with the details of each. You can see other examples in `stm32_min_dev`, `efr32_radio`, etc.

I think we should stick w/ that convention for boards. Each revision of a board might include significant changes, different chip onboard, etc. so having them be clearly defined per board version makes sense to me.

Now, since Zephyr doesn't include YAML files for the shields, we don't have much to use as a reference there, but I'd say useful to consider the current and future use cases.

* Non-split, single shield with compatibility with some module/board footprint.
* Keyboard with 2+ shields
* Keyboard with 2+ shields to be built in "dongle mode"
* Multi-shield keyboards where each component includes different additional hardware (e.g. left side only has encoder, right side only has OLED)
* Multiple revisions of each of the above.

To be honest... It really makes me wonder if we should be going the other way with this, and do a file per each uniquely named shield. Accept some duplication, but be *really* explicit.

The main detractor to that approach seems to be: "How do I know which pieces usually come together and I want to build as "one thing", if we go that route.

I wonder, if we go that route, w/ explicit per-shield file (e.g.. `lily58_left.yaml`/`lily58_right.yaml`, `lily58_pro_left.yaml`/`lily58_pro_right.yaml`), we then have some top level aggregation YAML file to create the connections, e.g. `lily58.yaml` or maybe `configs.yaml`.

That would have something like we talked about:

```yaml
configs:
  - identifier: lily58
    shields:
      - lily58_pro_left
      - lily58_pro_right
  - identifier: lily58_pro
    shields:
      - lily58_pro_left
      - lily58_pro_right
```

I believe the "per half" explicit docs are needed, in order to allow for each half including different hardware (encoders, OLEDs, etc) and our metadata should ideally capture that for future uses.

The `configs.yaml` file should make it easier to have tools like @joelspadin "user config tool" automate setup of a "full configuration"  from that as a starting point.

Non-split shields I think could *skip* the `config.yml` file, and the config tool could assume each shield should be built on their own.

@Nicell @innovaker Thoughts?

Reply:
> I wonder, if we go that route, w/ explicit per-shield file (e.g.. lily58_left.yaml/lily58_right.yaml, lily58_pro_left.yaml/lily58_pro_right.yaml), we then have some top level aggregation YAML file to create the connections, e.g. lily58.yaml or maybe configs.yaml.

+1 to this idea. I recommend going with `configs.yaml`. Having a consistently named file would make it much easier for a tool to identify the correct file. If a folder contains `lily58.yaml`, `lily58_pro.yaml`, `lily58_left.yaml`, `lily58_pro_left.yaml`, etc... then the only way I can think to pick out that `lily58.yaml` and `lily58_pro.yaml` are the top-level files would be to read each file.

> Non-split shields I think could skip the config.yml file, and the config tool could assume each shield should be built on their own.

This should work. While having a `configs.yaml` file *everywhere* would make it easier for people writing tools, it would also be pretty easy to check for the absence of such a file and then scan the directory for all .yaml files in that case.

Reply:
> 
> > Non-split shields I think could skip the config.yml file, and the config tool could assume each shield should be built on their own.
> 
> This should work. While having a `configs.yaml` file _everywhere_ would make it easier for people writing tools, it would also be pretty easy to check for the absence of such a file and then scan the directory for all .yaml files in that case.

Yeah, we'd probably need some rudimentary fallback, but I don't think it needs to be too fancy. We should be pushing for the `configs.yaml` file as the standard.

Reply:
I'm curious how this `configs.yaml` file would look for dongle mode. I know we don't really have a dongle mode yet to reference, but would we have a different shield variant on each shield for dongle mode?

As such:
```yaml
configs:
  - identifier: lily58
    shields:
      - lily58_pro_left
      - lily58_pro_right
  - identifier: lily58_pro
    shields:
      - lily58_pro_left
      - lily58_pro_right
  - identifier: lily58_dongle
    shields:
      - some_dongle
      - lily58_pro_left_dongle
      - lily58_pro_right_dongle
  - identifier: lily58_pro_dongle
    shields:
      - some_dongle
      - lily58_pro_left_dongle
      - lily58_pro_right_dongle
```

And then how do we handle the multiple possible dongles? Obviously, we could have more than one board that could work as a dongle.

Outside of this dongle question, this makes sense to me.

Reply:
I envisioned dongle mode being not part of this. You'd build all the shields for a version with overriding role to be peripheral, then build a standard dongle board + the keymap + # of dongles. I don't think the left/right sides change other than the config override.

So, long story short... I don't think it actually needs to be in the configs.

Reply:
Maybe obvious, but if we go with the configs.yaml file, we should probably give each config a `name` property to select the display name for splits. Tools could grab the name from the first listed board/shield (and probably should continue to use that as a fallback in case a name isn't given in configs.yaml), but that might pick out a name that has "left" or "right" in it.

Reply:
There's been some discussion related to this thread here:
https://discord.com/channels/719497620560543766/719544230497878116/821115603720077334

Reply:
Regarding Zephyr's conventions, I'm happy breaking with them for this issue.  As far as I can tell, the primary (only?) purpose of Zephyr's `${BOARD}.yaml` files is to feed the Twister test framework and there's hints that it's only a transitory solution:
- https://docs.zephyrproject.org/latest/guides/test/twister.html?highlight=twister#board-configuration
- https://docs.zephyrproject.org/latest/guides/dts/design.html?highlight=yaml#example-remaining-work

Whereas we're developing the metadata for these (immediate) purposes:
- web
  - hardware page
  - setup scripts and/or its replacement
- CI builds
- and maybe everything else here too: https://discord.com/channels/719497620560543766/719544230497878116/821417305081905163

So quite different purposes.  That said, there's always merit in cherry picking the best aspects of Zephyr's approach where beneficial.

Reply:
> The examples given above only seem to explore one type of grouping: left and right (halves)

I'd disagree with this. You can have however many parts in each shield array that you'd like. We just currently only have left and right halve examples.

Family, version, and variant seem like good classifiers. So a family would have version and versions would have variants? I think parts can probably be put in arrays as we have above.

Reply:
Are you talking about for use in the aggregate `configs.yaml` file? The individual files? Something else?

Reply:
If we're talking about the aggregate file, weld need to be clear on how we take that discrete taxonomy,.and still have an easy mapping to specific shield me for the build. If the convention is `family_rev_part_layout` with layout optional, I could see this working well.

Reply:
As I look at this more, and play w/ a JSON Schema to allow us to validate these files, I'm less and less convinced about the need to have the `parts` of each keyboard have their own file. It ends up with even *more* duplication, which I'm not sure is really necessary.

As I ponder the "each half has different hardware" part, I am seeing ways we can still do this without needing anything too crazy.

In particular, Looking at the JSON schema for what I'm playing with, you can easily end up w/ something like:

```
   "shield": {
      "title": "Shield",
      "type": "object",
      "additionalProperties": false,
      "required": [ "id", "name", "url", "type", "interconnects" ],
      "properties": {
        "id": { "$ref": "#/$defs/id" },
        "name": { "type": "string" },
        "url": { "type": "string", "format": "uri" },
        "manufacturer": { "type": "string" },
	"version": { "type": "string" },
        "type": {
          "type": "string",
	  "const": "shield"
        },
	"parts": { "$ref": "#/$defs/keyboard_parts" },
	"features": { "$ref": "#/$defs/features" },
        "interconnects": { "$ref": "#/$defs/interconnects" }
      }
```

Where you have a couple defs like:

```
    "keyboard_parts": {
      "type": "array",
      "items": {
        "oneOf": [
          { "type": "string" },
          { "$ref": "#/$defs/part_details" }
        ]
      }
    },
    "features": {
      "type": "array",
      "items": {
        "type": "string",
        "enum": [ "display", "encoder" ]
      }
    },
```

Which lets you have a basic set of features listed at the shield level, if those features apply to all parts, or you can specify the full details of the features present on each part of the shield/keyboard.

If we do this, we can get away with skipping the extra `configs.yml` file, and simplify this as "search for all *.zmk.yml files, inspect the content" to get each "build target" (which might include building multiple parts).

I'll post a full gist once I'm ready w/ sample files + schema to validate them.

Reply:
@petejohanson posted a gist here: https://discord.com/channels/719497620560543766/719544230497878116/825361233074454529

Reply:
Thanks @petejohanson!

---

As we work through this, let's keep in mind that each purpose I listed has quite different needs.  Zephyr only cares about a unique id, whereas everything else we're building cares about the attributes and relationships, the more meaningful the better.  @petejohanson's exploration of a schema - or more broadly speaking the underlying data model - will help us define these aspects.

---

I'd like to highlight that `version` and `variant` (and `part` to some extent) are sometimes defined by the maker.   Each of them are optional and independent, and they don't always follow the hierarchy/relationships we might expect.

---

> Are you talking about for use in the aggregate configs.yaml file? The individual files? Something else?

I was talking conceptually - data modelling - rather than files.

---

I think listing the features is a good idea.  I'd been considering it a future effort, but I can see the logic in defining them as part of the initial design.

---

`part` currently conveys a few meanings such as:
- which bit?
- what part of the body (hand) uses it?
  - and coupled with that: probable physical position relative to the user and maybe the other parts too.
- what role?
  - This is debatable because it's still a largely undeveloped grey area.
  - It's an important concept but I think we should fully explore other topologies before defining it.

So perhaps we want to consider capturing these meanings individually?  They may not be critical now, but they probably will be in the future.

Reply:
I'm putting some parts of the example schema @petejohanson has made in here to offer suggestions/thoughts.

> ```
>     "keyboard_parts": {
>       "type": "array",
>       "items": {
>         "oneOf": [
>           { "type": "string" },
>           { "$ref": "#/$defs/part_details" }
>         ]
>       }
>     },
> ```

I think we need a better property name than this since, although we are a keyboard firmware, there's the opportunity to have lots more input devices added in the future. Parts is too generic. As is components or nodes. Sub-assemblies makes more sense to me, but it's a bit long-winded and still doesn't feel right.

Along with this, what does "type" mean? I'm assuming it's the `id` associated with this part of the network, but we're still lacking a lot of other information about what this part actually is/does as @innovaker pointed out. Although is it truly important for us to deliberately state what this part does? Maybe it could be optional descriptors.

> ```
>     "variant": {
>       "oneOf": [
> 	{ "type": "string" },
> 	{
>           "type": "object",
> 	  "required": [ "id", "features" ],
>           "properties": {
>             "id": { "$ref": "#/$defs/id" },
>             "features": { "$ref": "#/$defs/features" }
>           }
> 	}
>       ]
>     },
> ```

I think it makes sense to have the `id`, which is associated with the variant overlay, but I'm not sure how features would work here. Are they adding on to existing features or overwriting the "parent" features. This seems like a tricky piece, as variants could have the exact same features, added features, or a whole different set of features. How do we make this easy to manage for people writing a variant?

> ```
>     "features": {
>       "type": "array",
>       "items": {
>         "type": "string",
>         "enum": [ "display", "encoder" ]
>       }
>     },
> ```

I'm glad this section exists, but I'm wondering how we keep this up to date and allow for new features to be added. We definitely don't want this to be opened up because then it becomes a nightmare for filtering and any data processing of features. We need a simple way to define and add features without having to adjust the schema directly everytime IMO.

> ```
>     "interconnects": {
>       "type": "array",
>       "minItems": 1,
>       "items": {
>         "type": "string",
>         "minLength": 1
>       }
>     },
> ```

After thinking about it for a while, I really do think a simple interconnect array is sufficient for our needs of `shield <-> board` connection because of how Zephyr works. I'm wondering though how we'll do the translation from `id` to a readable name as well as how we'll moderate the `id`s of generic interconnects that aren't specific to boards i.e. `pro_micro` (generic) vs. `proton_c` (board specific).

Outside of these parts, the board/shield properties look relatively straightforward outside of the larger discussion about variants and versions. Thanks for putting this initial schema together @petejohanson !